Double allocation error of FIELD variable [w3wavemd]

ukmo-ccbunney opened the pull request:
Relates to issue https://github.com/NOAA-EMC/WW3/issues/252

When running on an `SMC` grid using `SHRD` and `OMPG`/`OMPX` switches, `ww3_shel` crashes in the `w3wavemd` module due to an array being allocated twice.

The issue arises due to some incorrect logic relating to the allocation of the `FIELD` variable. There is a defunct `OMP PARALLEL` section starting at line 1731 that was disabled some time ago in PR https://github.com/NOAA-EMC/WW3/pull/160 due to it causing bit-for-bit differences. However, the logic for allocating the FILED array as a private variable inside the parallel section (rather than a shared variable outside) was left in - and this logic is incorrect. When the model is run with a combination of `SMC` and `OMPX` switches, the array is erroneously allocated twice.

The simplest solution is to keep only the initial allocation at line 580 - 583 and remove the allocation in the parallel section. Furthermore, as the `OMPX` statements are now defunct, we can remove them and the associated `FLOMP` logical. This also requires only a single `DEALLOCATE` statement.

_NOTE: If it is considered that the OMP section starting at line 1731 might be re-instated in the future, I can keep the second allocation and fix the incorrect logic._

ukmo-ccbunney committed:
Fixed allocation of FIELD variable.
Also removed redundant OMPX switches.

ukmo-ccbunney committed:
Removed defunct OMPX switches and FLOMP variable

ukmo-ccbunney commented:
A selection of regtests show no unexpected differences (ww3_tp2.18 difference in ww3_prtide_current.out is an existing issue).
[matrixCompSummary_bf_smc_omp.zip](https://github.com/ukmo-waves/WW3/files/5225002/matrixCompSummary_bf_smc_omp.zip)

ukmo-jianguo-li commented:
The simplest way to solve this conflict is to comment off the following line in w3wavemd.ftn:

!/OMPX      FLOMP  = .TRUE.

If OMPX option is no longer required, tiding up the code by removing all OMPX related lines will also correct the bug.

One question is that why the OMPX option is still tested if it is no longer used?

ukmo-ccbunney commented:
Thanks @ukmo-jianguo-li. In response:

> The simplest way to solve this conflict is to comment off the following line in w3wavemd.ftn: !/OMPX FLOMP = .TRUE. 

I've done more that that - I've removed it and all other lines relating to OMPX and FLOMP. I've also removed the second ALLOCATE statement in the OMP block.

> If OMPX option is no longer required, tiding up the code by removing all OMPX related lines will also correct the bug. 

Yup - that's what I've done. :)

>One question is that why the OMPX option is still tested if it is no longer used?

It's true that the w3wavemd module was the only module using the OMPX switch. I have not removed the switch from the list of valid switches as it is likely that future development might make use of it. 

Are you happy with the changes I have made?

ukmo-jianguo-li commented:
The changes are ok as long as no one complains about removing the OMPX option.  As most of the OMPX lines have been commented off in the old code, it seems fine to go ahead with your changes.

ukmo-jianguo-li reviewed (approved):
See comments under the conversation tab.
